File: javaProject/src/ui/ConsoleMenu.java
package ui;

import domain.DirectedGraph;
import domain.Edge;
import service.GraphIO;
import service.RandomGraphMaker;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.stream.Collectors;

public class ConsoleMenu {
    private final Scanner scanner;
    private DirectedGraph graph;
    private DirectedGraph backup;
    private boolean hasBackup;

    public ConsoleMenu(DirectedGraph graph) {
        this.graph = graph;
        this.backup = new DirectedGraph(0);
        this.hasBackup = false;
        this.scanner = new Scanner(System.in);
    }

    public void run() {
        while (true) {
            printMenu();
            String option = readString("Choose an option: ");

            if ("0".equals(option)) {
                System.out.println("Exiting application.");
                return;
            }

            try {
                handleOption(option);
            } catch (Exception error) {
                System.out.println("Error: " + error.getMessage());
            }

            System.out.println();
        }
    }

    private void printMenu() {
        System.out.println("Directed Graph Console");
        System.out.println("1. Load graph from file");
        System.out.println("2. Save graph to file");
        System.out.println("3. Generate random graph");
        System.out.println("4. Show graph summary");
        System.out.println("5. List all vertices");
        System.out.println("6. Inspect a vertex");
        System.out.println("7. Check whether an edge exists");
        System.out.println("8. View an edge cost");
        System.out.println("9. Update an edge cost");
        System.out.println("10. Add a vertex");
        System.out.println("11. Remove a vertex");
        System.out.println("12. Add an edge");
        System.out.println("13. Remove an edge");
        System.out.println("14. Copy current graph to backup");
        System.out.println("15. Restore graph from backup copy");
        System.out.println("16. Print all edges");
        System.out.println("0. Exit");
    }

    private void handleOption(String option) throws IOException {
        switch (option) {
            case "1" -> loadGraph();
            case "2" -> saveGraph();
            case "3" -> generateRandomGraph();
            case "4" -> showSummary();
            case "5" -> listVertices();
            case "6" -> inspectVertex();
            case "7" -> checkEdge();
            case "8" -> showEdgeCost();
            case "9" -> updateEdgeCost();
            case "10" -> addVertex();
            case "11" -> removeVertex();
            case "12" -> addEdge();
            case "13" -> removeEdge();
            case "14" -> backupGraph();
            case "15" -> restoreGraph();
            case "16" -> printEdges();
            default -> throw new IllegalArgumentException("Unknown menu option");
        }
    }

    private void loadGraph() throws IOException {
        String fileName = readString("File name: ");
        graph = GraphIO.readGraph(fileName);
        System.out.println("Loaded graph from " + fileName + ".");
    }

    private void saveGraph() throws IOException {
        String fileName = readString("File name: ");
        GraphIO.writeGraph(fileName, graph);
        System.out.println("Saved graph to " + fileName + ".");
    }

    private void generateRandomGraph() {
        int vertices = readInt("Number of vertices: ");
        int edges = readInt("Number of edges: ");
        int minCost = readInt("Minimum edge cost: ");
        int maxCost = readInt("Maximum edge cost: ");

        graph = RandomGraphMaker.makeRandomGraph(vertices, edges, minCost, maxCost);
        System.out.println("Random graph generated.");
    }

    private void showSummary() {
        System.out.println("Vertices: " + graph.getVertices());
        System.out.println("Edges: " + graph.getEdges());
    }

    private void listVertices() {
        List<Integer> vertices = graph.parseVertices();
        if (vertices.isEmpty()) {
            System.out.println("The graph has no vertices.");
            return;
        }

        System.out.println("Vertices: " + formatCollection(vertices));
    }

    private void inspectVertex() {
        int vertex = readInt("Vertex: ");
        List<Integer> inbound = graph.parseInboundNeighbors(vertex);
        List<Integer> outbound = graph.parseOutboundNeighbors(vertex);

        System.out.println("In-degree: " + graph.getInDegree(vertex));
        System.out.println("Out-degree: " + graph.getOutDegree(vertex));
        System.out.println("Inbound neighbors: " + formatCollection(inbound));
        System.out.println("Outbound neighbors: " + formatCollection(outbound));
    }

    private void checkEdge() {
        Edge edge = readEdge();
        boolean exists = graph.isEdge(edge.getStart(), edge.getEnd());
        System.out.println(
                "Edge " + edge.getStart() + " -> " + edge.getEnd() + " exists: " + (exists ? "yes" : "no")
        );
    }

    private void showEdgeCost() {
        Edge edge = readEdge();
        System.out.println(
                "Cost of " + edge.getStart() + " -> " + edge.getEnd() + ": "
                        + graph.getEdgeCost(edge.getStart(), edge.getEnd())
        );
    }

    private void updateEdgeCost() {
        Edge edge = readEdge();
        int cost = readInt("New cost: ");
        graph.setEdgeCost(edge.getStart(), edge.getEnd(), cost);
        System.out.println("Edge cost updated.");
    }

    private void addVertex() {
        int vertex = graph.addVertex();
        System.out.println("Added vertex " + vertex + ".");
    }

    private void removeVertex() {
        int vertex = readInt("Vertex to remove: ");
        graph.removeVertex(vertex);
        System.out.println("Removed vertex " + vertex + ".");
    }

    private void addEdge() {
        Edge edge = readEdge();
        int cost = readInt("Cost: ");
        graph.addEdge(edge.getStart(), edge.getEnd(), cost);
        System.out.println("Added edge " + edge.getStart() + " -> " + edge.getEnd() + ".");
    }

    private void removeEdge() {
        Edge edge = readEdge();
        graph.removeEdge(edge.getStart(), edge.getEnd());
        System.out.println("Removed edge " + edge.getStart() + " -> " + edge.getEnd() + ".");
    }

    private void backupGraph() {
        backup = graph.copyGraph();
        hasBackup = true;
        System.out.println("Backup copy created.");
    }

    private void restoreGraph() {
        if (!hasBackup) {
            throw new IllegalStateException("No backup graph is available");
        }

        graph = backup.copyGraph();
        System.out.println("Graph restored from backup.");
    }

    private void printEdges() {
        Map<Edge, Integer> costs = graph.getCosts();
        if (costs.isEmpty()) {
            System.out.println("The graph has no edges.");
            return;
        }

        for (Map.Entry<Edge, Integer> entry : costs.entrySet()) {
            Edge edge = entry.getKey();
            System.out.println(edge.getStart() + " -> " + edge.getEnd() + " (cost = " + entry.getValue() + ")");
        }
    }

    private int readInt(String prompt) {
        System.out.print(prompt);

        if (!scanner.hasNext()) {
            throw new IllegalStateException("No more input available");
        }

        String value = scanner.next();
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException error) {
            throw new IllegalArgumentException("Invalid integer input", error);
        }
    }

    private String readString(String prompt) {
        System.out.print(prompt);

        if (!scanner.hasNext()) {
            throw new IllegalStateException("No more input available");
        }

        return scanner.next();
    }

    private Edge readEdge() {
        int u = readInt("Start vertex: ");
        int v = readInt("End vertex: ");
        return new Edge(u, v);
    }

    private String formatCollection(List<Integer> values) {
        if (values.isEmpty()) {
            return "none";
        }

        return values.stream()
                .map(String::valueOf)
                .collect(Collectors.joining(" "));
    }
}

File: pythonProject/src/ui/consoleMenu.py
from __future__ import annotations

from src.domain.graph import UndirectedGraph
from src.service.components import connectedComponentsBfs
from src.service.graphIo import readGraph, writeGraph
from src.service.randomGraph import makeRandomGraph


class ConsoleMenu:
    def __init__(self, graph: UndirectedGraph | None = None):
        self._graph = graph if graph is not None else UndirectedGraph()
        self._backup: UndirectedGraph | None = None

    def run(self) -> None:
        while True:
            self._printMenu()
            option = input("Choose an option: ").strip()

            if option == "0":
                print("Exiting application.")
                return

            try:
                self._handleOption(option)
            except ValueError as error:
                print(f"Error: {error}")
            except FileNotFoundError as error:
                print(f"File error: {error}")

            print()

    def _printMenu(self) -> None:
        print("Undirected Graph Console")
        print("1. Load graph from file")
        print("2. Save graph to file")
        print("3. Generate random graph")
        print("4. Show graph summary")
        print("5. List all vertices")
        print("6. Inspect a vertex")
        print("7. Check whether an edge exists")
        print("8. View an edge cost")
        print("9. Update an edge cost")
        print("10. Add a vertex")
        print("11. Remove a vertex")
        print("12. Add an edge")
        print("13. Remove an edge")
        print("14. Copy current graph to backup")
        print("15. Restore graph from backup copy")
        print("16. Print all edges")
        print("17. Connected components by BFS")
        print("0. Exit")

    def _handleOption(self, option: str) -> None:
        actions = {
            "1": self._loadGraph,
            "2": self._saveGraph,
            "3": self._generateRandomGraph,
            "4": self._showSummary,
            "5": self._listVertices,
            "6": self._inspectVertex,
            "7": self._checkEdge,
            "8": self._showEdgeCost,
            "9": self._updateEdgeCost,
            "10": self._addVertex,
            "11": self._removeVertex,
            "12": self._addEdge,
            "13": self._removeEdge,
            "14": self._backupGraph,
            "15": self._restoreGraph,
            "16": self._printEdges,
            "17": self._runComponentsBfs,
        }

        action = actions.get(option)
        if action is None:
            raise ValueError("Unknown menu option")

        action()

    def _loadGraph(self) -> None:
        fileName = input("File name: ").strip()
        self._graph = readGraph(fileName)
        print(f"Loaded undirected graph from {fileName}.")

    def _saveGraph(self) -> None:
        fileName = input("File name: ").strip()
        writeGraph(fileName, self._graph)
        print(f"Saved graph to {fileName}.")

    def _generateRandomGraph(self) -> None:
        vertices = self._readInt("Number of vertices: ")
        edges = self._readInt("Number of edges: ")
        minCost = self._readInt("Minimum edge cost: ")
        maxCost = self._readInt("Maximum edge cost: ")
        self._graph = makeRandomGraph(vertices, edges, minCost, maxCost)
        print("Random graph generated.")

    def _showSummary(self) -> None:
        print("Type: undirected")
        print(f"Vertices: {self._graph.vertexCount()}")
        print(f"Edges: {self._graph.edgeCount()}")

    def _listVertices(self) -> None:
        vertices = self._graph.parseVertices()
        if not vertices:
            print("The graph has no vertices.")
            return

        print("Vertices:", " ".join(str(vertex) for vertex in vertices))

    def _inspectVertex(self) -> None:
        vertex = self._readInt("Vertex: ")
        neighbors = self._graph.parseNeighbors(vertex)
        print(f"Degree: {self._graph.degree(vertex)}")
        print("Neighbors:", self._formatCollection(neighbors))

    def _checkEdge(self) -> None:
        u, v = self._readEdge()
        exists = self._graph.isEdge(u, v)
        print(f"Edge {u} -- {v} exists: {'yes' if exists else 'no'}")

    def _showEdgeCost(self) -> None:
        u, v = self._readEdge()
        print(f"Edge cost: {self._graph.getEdgeCost(u, v)}")

    def _updateEdgeCost(self) -> None:
        u, v = self._readEdge()
        cost = self._readInt("New cost: ")
        self._graph.setEdgeCost(u, v, cost)
        print("Edge cost updated.")

    def _addVertex(self) -> None:
        answer = input("Vertex id (leave empty for automatic): ").strip()
        vertex = self._graph.addVertex(None if answer == "" else int(answer))
        print(f"Added vertex {vertex}.")

    def _removeVertex(self) -> None:
        vertex = self._readInt("Vertex to remove: ")
        self._graph.removeVertex(vertex)
        print(f"Removed vertex {vertex}.")

    def _addEdge(self) -> None:
        u, v = self._readEdge()
        cost = self._readInt("Cost: ")
        self._graph.addEdge(u, v, cost)
        print("Edge added.")

    def _removeEdge(self) -> None:
        u, v = self._readEdge()
        self._graph.removeEdge(u, v)
        print("Edge removed.")

    def _backupGraph(self) -> None:
        self._backup = self._graph.copyGraph()
        print("Backup copy created.")

    def _restoreGraph(self) -> None:
        if self._backup is None:
            raise ValueError("No backup graph is available")

        self._graph = self._backup.copyGraph()
        print("Graph restored from backup.")

    def _printEdges(self) -> None:
        edges = self._graph.parseEdges()
        if not edges:
            print("The graph has no edges.")
            return

        for (u, v), cost in edges:
            print(f"{u} -- {v} (cost = {cost})")

    def _runComponentsBfs(self) -> None:
        self._printComponents(connectedComponentsBfs(self._graph), "BFS")

    @staticmethod
    def _readInt(prompt: str) -> int:
        return int(input(prompt).strip())

    def _readEdge(self) -> tuple[int, int]:
        u = self._readInt("Start vertex: ")
        v = self._readInt("End vertex: ")
        return u, v

    @staticmethod
    def _formatCollection(values: list[int]) -> str:
        if not values:
            return "none"
        return " ".join(str(value) for value in values)

    @staticmethod
    def _printComponents(components: list[UndirectedGraph], method: str) -> None:
        print(f"Connected components found by {method}: {len(components)}")
        for index, component in enumerate(components, start=1):
            print(
                f"{index}. vertices = {component.parseVertices()}, "
                f"edges = {component.edgeCount()}"
            )
